src/Services/ContentNormalizer.php

```diff
@@ -18,37 +19,32 @@ public function normalize(string $content): string
 
     public function trimLength(string $content, int $maxLength, bool $onWords = true): string
     {
-        return $onWords ? $this->wordSplitter($content, $maxLength) : trim(mb_substr($content, 0, $maxLength));
+        $str = new UnicodeString($content);
+
+        return $onWords ? $this->wordSplitter($str, $maxLength) : $str->truncate($maxLength)->trim()->toString();
     }
 
     /**
      * Trim content length on whole words.
      *
-     * @see https://stackoverflow.com/questions/79960/how-to-truncate-a-string-in-php-to-the-word-closest-to-a-certain-number-of-chara
-     *
-     * @param string $content
+     * @param UnicodeString $str
      *   The content to trim in length
      * @param int $maxLength
      *   The max length of content to return
      *
      * @return string
      *   The trimmed content wit max length
      */
-    private function wordSplitter(string $content, int $maxLength): string
+    private function wordSplitter(UnicodeString $str, int $maxLength): string
     {
-        $parts = preg_split('/([\s\n\r]+)/u', $content, -1, PREG_SPLIT_DELIM_CAPTURE);
-        $parts_count = count($parts);
-
-        $length = 0;
-        $last_part = 0;
-        for (; $last_part < $parts_count; ++$last_part) {
-            $currentPartLength = mb_strlen($parts[$last_part]);
-            if ($length + $currentPartLength > $maxLength) {
-                break;
-            }
-            $length += $currentPartLength;
+        $str = $str->truncate($maxLength, '', false)->trim();
+
+        if ($str->length() > $maxLength) {
+            // As the truncate function above, have returned a string longer then max length. Remove the last word from
+            // the string to ensure that it is below max length.
+            return preg_replace('/\W\w+\s*(\W*)$/', '$1', $str->toString());
```

Comment:
This will not work if the string ends with a non-word character. You must also handle unicode in your regex:

```suggestion
        if ($str->length() > $maxLength) {
            // The truncate function above has returned a string longer then max length. Remove the last space character and any non-space chacters from end of string.
            // the string.
            $str = new UnicodeString(preg_replace('/\s+[^\s]+$/u', '', $str->toString()));
            // Truncate the string hard if the string is still longar than the max length, e.g. if it's a single very long word.
            if ($str->length() > $maxLength) {
                $str = $str->truncate($maxLength);
            }
        }
```

We need a test for this function!